kroxylicious/src/test/java/io/kroxylicious/proxy/internal/ResponseOrdererTest.java

```diff
@@ -0,0 +1,212 @@
+        whenWriteOutboundMessage(response);
+        thenOrdererStateEquals(0, 0);
+        thenOutboundContains(response);
+    }
+
+    @Test
+    void testMultipleInflightMessagesThatRespondInOrder() {
+        whenWriteInboundMessage(new TestFrame(1));
+        whenWriteInboundMessage(new TestFrame(2));
+        thenOrdererStateEquals(2, 0);
+        TestFrame response = new TestFrame(1);
+        whenWriteOutboundMessage(response);
+        thenOrdererStateEquals(1, 0);
+        thenOutboundContains(response);
+        TestFrame response2 = new TestFrame(2);
+        whenWriteOutboundMessage(response2);
+        thenOrdererStateEquals(0, 0);
+        thenOutboundContains(response2);
+    }
+
+    @Test
+    void testMultipleInflightMessagesThatRespondOutOfOrder() {
+        whenWriteInboundMessage(new TestFrame(1));
+        whenWriteInboundMessage(new TestFrame(2));
+        thenOrdererStateEquals(2, 0);
+        TestFrame response = new TestFrame(2);
+        whenWriteOutboundMessage(response);
+        thenOrdererStateEquals(2, 1);
+        thenOutboundIsEmpty();
+        TestFrame response2 = new TestFrame(1);
+        whenWriteOutboundMessage(response2);
+        thenOrdererStateEquals(0, 0);
+        thenOutboundContains(response2);
+        thenOutboundContains(response);
+    }
+
+    @Test
+    void testComplexScenarioWherePendingResponsesCanBePartiallyDrained() {
+        whenWriteInboundMessage(new TestFrame(1));
+        whenWriteInboundMessage(new TestFrame(2));
+        whenWriteInboundMessage(new TestFrame(3));
+        whenWriteInboundMessage(new TestFrame(4));
+        thenOrdererStateEquals(4, 0);
+        TestFrame response = new TestFrame(2);
+        whenWriteOutboundMessage(response);
+        thenOrdererStateEquals(4, 1);
+        thenOutboundIsEmpty();
+
+        TestFrame response2 = new TestFrame(4);
+        whenWriteOutboundMessage(response2);
+        thenOrdererStateEquals(4, 2);
+        thenOutboundIsEmpty();
+
+        TestFrame response3 = new TestFrame(1);
+        whenWriteOutboundMessage(response3);
+        // the response for correlationId 2 should be drained after the response for correlationId 1 is written
+        // but the response for correlationId 4 cannot be sent until we have written the response for correlationId 3
+        thenOrdererStateEquals(2, 1);
+        thenOutboundContains(response3);
+        thenOutboundContains(response);
+        thenOutboundIsEmpty();
+
+        TestFrame response4 = new TestFrame(3);
+        whenWriteOutboundMessage(response4);
+        thenOrdererStateEquals(0, 0);
+        thenOutboundContains(response4);
+        thenOutboundContains(response2);
+        thenOutboundIsEmpty();
+    }
+
+    @Test
+    void testMultiplePendingAreDrained() {
+        whenWriteInboundMessage(new TestFrame(1));
+        whenWriteInboundMessage(new TestFrame(2));
+        whenWriteInboundMessage(new TestFrame(3));
+        thenOrdererStateEquals(3, 0);
```

Comment:
Do we need this assertion? Given we have other tests that have confirmed this state?